src/rt.rs

```diff
@@ -5,8 +5,43 @@
 //! If the `runtime` feature is disabled, the types in this module can be used
 //! to plug in other runtimes.
 
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures_core::Future;
+
 /// An executor of futures.
 pub trait Executor<Fut> {
     /// Place the future into the executor to be run.
     fn execute(&self, fut: Fut);
 }
+
+/// A timer which provides timer-like functions, similar to tokio::time::*.
+pub trait Timer {
+    /// An analogue of tokio::time::sleep.
+    fn sleep(&self, duration: Duration) -> Box<dyn Sleep + Unpin>;
+
+    /// An analogue of tokio::time::sleep_until.
+    fn sleep_until(&self, deadline: Instant) -> Box<dyn Sleep + Unpin>;
+}
+
+/// The generic version of tokio::time::Sleep, which itself is the output of tokio::time::sleep
+pub trait Sleep: Send + Sync + Unpin + Future<Output = ()> {
+    /// An analogue of tokio::time::Sleep::deadline.
+    fn deadline(&self) -> Instant;
+
+    /// An analogue of tokio::time::Sleep::reset.
+    fn reset(self: Pin<&mut Self>, deadline: Instant);
```

Comment:
[Continuing the conversation](https://github.com/hyperium/hyper/pull/2904#discussion_r957859176), where @dhobsd said:

> Thanks for remembering us and reaching out! (Sorry for delay!)
> 
> Fuchsia's async executor uses deadlines for its time, so that's at least good. Two points to bring up:
> 
> - reset in tokio says "This function can be called both before and after the future has completed." In [fuchsia_async's Timer](https://fuchsia-docs.firebaseapp.com/rust/fuchsia_async/struct.Timer.html), this isn't true -- you can only reset a timer that's already fired. I'm also not sure what it means to reset a timer that hasn't fired because it's hard to guarantee you won't observe the signal. Does hyper reset timers that haven't fired?
> - is_elapsed says that it returns true when the time has elapsed, but is this the same as saying the timer has already fired?
> 
> I guess a third question: will hyper rely on these divergent behaviors? If not, perhaps we can make the documentation stricter in hyper's trait?

Reply:
So I looked around how we use these methods, and here's what I've found:

- `reset` is used for the HTTP/1 header read timeout, so that successive reads simply move the timer further out, instead of allocating a new one.
- Doesn't look like we use the other methods, so I've been happy to toss them.

The `reset` method seems to just be to allow for optimizations, but we _could_ simply provide a default implementation of the method somehow... Basically, if it didn't exist, we'd just call `timer.sleep(next_dur)` and replace the old one. Hm, that might mean the method needs to be on `Timer` instead, since we can't make a default implementation that creates a new sleep without the `Timer`. Maybe then a `timer.reset(sleep, next_dur)`.

We even could just leave it off entirely to start with, since adding a method with a default implementation is fine. It would just mean potentially unnecessary allocations when using Tokio.